File: java-files/Problem41.java
package org.algorithms.euler;

/**
 * Pandigital prime
 * 
 * Problem 41
 * We shall say that an n-digit number is pandigital if it makes use of all the
 * digits 1 to n exactly once. For example, 2143 is a 4-digit pandigital and is
 * also prime.
 * 
 * What is the largest n-digit pandigital prime that exists?
 * 
 * Solution
 * 
 * Start from the largest number that is posible for this case 999,999,999
 * 
 * Pandigital that is prime: 7652413
 * 
 */
public class Problem41 {
  
  public static boolean isPandigital(String s) {
    if (s.length() == 0) {
      return false;
    }
    
    for (int i = 0; i < s.length(); i++) {
      if (!s.contains(String.valueOf(i + 1))) {
        return false;
      }
    }
    return true;
  }
  
  
  public static void main(String[] args) {
    int size = 1000000000;
    boolean sieve[] = SieveOfEratosthenes.create(size);
    System.out.println("Sieve is ready!");
    //SieveOfEratosthenes.print(sieve);
    
    for (int i = size -1; i >= 0; i--) {
      if (sieve[i]) {
        if (isPandigital(String.valueOf(i))) {
          System.out.println("Pandigital that is prime: " + i);
          return;
        }
      }
    }
    
  }
}

File: c-files/problem_41/problem41.c
/*
 * Pandigital prime
 *
 * Problem 41
 *
 * We shall say that an n-digit number is pandigital if it makes use of all the digits 1 to n
 * exactly once. For example, 2143 is a 4-digit pandigital and is also prime.
 *
 * What is the largest n-digit pandigital prime that exists?
 *
 *  Created on: Feb 20, 2014
 *      Author: jcyescas
 */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../lib/util.h"

int main() {
  printf("\nPandigital Prime\n");

  size_t size = 987654322;
  bool *sieve = malloc(sizeof(bool) * size);
  sieve_eratosthenes_create(sieve, size);

  char value[10];
  int i;
  for (i = size - 1; i >=0; i--) {
    if (sieve[i]) {
      sprintf(value, "%d", i);
      if(is_pandigital(value)) {
        printf("\nThe largest n-digit pandigital prime that exists is %d", i);
        break;
      }
    }
  }

  free(sieve);
  return 0;
}

